builtin/mainmenu/tab_online.lua

```diff
@@ -210,21 +210,66 @@ local function search_server_list(input)
 	for i = 1, #serverlistmgr.servers do
 		local server = serverlistmgr.servers[i]
 		local found = 0
+		local sername_keywords = {}
+		local sername_keywords_matchtimes = {}
+		local server_desc_keywords = {}
+		local server_desc_keywords_matchtimes = {}
+		-- Make search precise for exactly matched server
+		-- Or at least more exactly matched server should be rated higher at the match scoring
+		if server.name then
+			for word in server.name:gmatch("%S+") do
+				word = word:gsub("(%W)", "%%%1"):lower()
+				table.insert(sername_keywords, word)
+				if sername_keywords_matchtimes[word] == nil then
+					sername_keywords_matchtimes[word] = 1
+				else
+					sername_keywords_matchtimes[word] = sername_keywords_matchtimes[word] + 1
+				end
+			end
+		end
+		if server.description then
+			for word in server.description:gmatch("%S+") do
+				word = word:gsub("(%W)", "%%%1"):lower()
+				table.insert(server_desc_keywords, word)
+				if server_desc_keywords_matchtimes[word] == nil then
+					server_desc_keywords_matchtimes[word] = 1
+				else
+					server_desc_keywords_matchtimes[word] = server_desc_keywords_matchtimes[word] + 1
+				end
+			end
+		end
 		for k = 1, #keywords do
-			local keyword = keywords[k]
-			if server.name then
-				local sername = server.name:lower()
-				local _, count = sername:gsub(keyword, keyword)
-				found = found + count * 4
+			local keyword
+			local sername_keyword
+			local desc_keyword
+			if keywords[k] ~= nil then
+				keyword = keywords[k]:lower()
+			else
+				goto continue
 			end
-
-			if server.description then
-				local desc = server.description:lower()
-				local _, count = desc:gsub(keyword, keyword)
-				found = found + count * 2
+			if sername_keywords[k] ~= nil then
+				sername_keyword = sername_keywords[k]
+			end
+			if server_desc_keywords[k] ~= nil then
+				desc_keyword = server_desc_keywords[k]
+			end
+			if keyword == sername_keyword then
```

Comment:
I don't really follow the logic here. You're comparing the k-th keyword to the k-th server name keyword. If they match, you increase the `found` score.

This is neither `AND`ing keywords nor is it `OR`ing and weighing them.